app/controllers/master/providers/plans_controller.rb

```diff
@@ -8,6 +8,9 @@ def edit
   end
 
   def update
+    authorize! :update, :provider_plans
+    raise CanCan::AccessDenied unless current_user.has_access_to_service?(@new_plan.issuer_id)
```

Comment:
why not `authorize! :show, @new_plan.issuer` ?

Reply:
@macejmic you also need to keep in mind those that have `nil` meaning that they have access. If they didn't have access they would have `'[]'`.
So it should be for the service in particular using `user.has_access_to_service?(service)`.

This is one of the 2 things that I mentioned in `Notes for the reviewers`, please read that section. It can be done and it is not difficult but it is not trivial, specially from the view `show.html`.

My point in `notes for the reviewers` is: should I do these 2 things now or should I create a Jira issue for them? It depends on how much time we have 😄 👍

Reply:
Thinking of that, (for later) I really like to use cancan instead of having the `#has_access_to_service?` method

```ruby
authorize! :update, :provider_plans
authorize! :manage, Service, @new_plan.issuer_id
```

`user.accessible_services.where_values_hash` will return a value (Service.match_all) that will always match all services instead of `nil` so it can be used in cancan

Reply:
@hallelujah @macejmic 
I agree with using this (well, kinda, I have another idea that I will mention later in the next comment):
```ruby
authorize! :update, :provider_plans
authorize! :manage, Service, @new_plan.issuer_id
```

My point here is that defining the ability, we shouldn't do this:
```ruby
can [:read, :show, :edit, :update], Service, user.accessible_services.where_values_hash
```
but instead:
```ruby
can :manage, Service do |service|
  user.has_access_to_service?(service)
end
```

The reason of this is because the way we save `member_permission_service_ids` in DB is the following:
- `nil` means allow them all
- `[]` means do not allow any
- `[1, 2, 3, 4]` means allowing services with ids 1,2,3,4

So if we apply this `hash-of-conditions`, it will not work for the case of `nil` 😄

Reply:
And what I said that I have another idea that I would mention in this comment is, instead of having `authorize! :update, :provider_plans` and `authorize! :manage, Service, @new_plan.issuer_id`, having directly `authorize! :update, @cinstance` 

And then in the ability:
```
if user.has_permission?(:partners)
  can :manage, Application do |cinstance|
    user.has_access_to_service?(cinstance.service_id)
  end
end
```

Reply:
Check this: 😄 
https://github.com/3scale/porta/blob/99ec8db6b7bd4b8286034cb212b4a1ecc6c5b016/app/models/user/permissions.rb#L87-L96

Reply:
And also see this (specially the comment):
https://github.com/3scale/porta/blob/99ec8db6b7bd4b8286034cb212b4a1ecc6c5b016/app/models/user/permissions.rb#L74-L81

Reply:
all that is true, but looking at: https://github.com/3scale/porta/blob/master/app/models/user.rb#L174-L176

along with: https://github.com/3scale/porta/blob/master/app/models/service.rb#L100-L110

it looks like the `nil` shouldn't be the case..

would be super awesome to have a test for that, to be completely sure..

Reply:
I will check and get back to you 👍 💪

Reply:
Ahaaa, you are right @macejmic 👍 Thanks! 💪 
Because `forbidden_some_services?` is checking this:
https://github.com/3scale/porta/blob/99ec8db6b7bd4b8286034cb212b4a1ecc6c5b016/app/models/user/permissions.rb#L98-L100

About the test, you are absolutely right too 👍 I added what it was missing imo in:
https://github.com/3scale/porta/pull/816